Metrics Webapp - Population Index- Final Version

Wahaj-Ali opened the pull request:
### Highlights
- It is a SPA built with React and Redux.
- It uses [API](https://restcountries.com/#api-endpoints-v3) to fetch the data(World countries and their details).
- It contains a "main page" with a list of all the countries with some details and a "Detail page"with all the details about a particular country.
- Each page has a unique path.
- The project is deployed using Netlify. The link to the deployed version is in the Readme file.
- Link to the video presentation is [here](https://www.loom.com/share/7d0c87edbe014e3d989e7f2df117e94c).

Wahaj-Ali committed:
Install react-test-renderer

In this commit
- Install react-test-renderer.
- Edit package.json to test files containing axios.

Wahaj-Ali committed:
Create Header.test.js

Create test for header component using react-testing-library.

Wahaj-Ali committed:
Create test for CountryList component

In this commit
- Add test for CountryList component.

Wahaj-Ali committed:
Add test for API

In this commit
- Add test for "fetchApi"data.
- Remove linter errors.

Wahaj-Ali committed:
Set up test for Dashboard page.

Wahaj-Ali committed:
Update Dashboard.test.js

Remove eslint errors.

Wahaj-Ali committed:
Improved styling to match the design guidelines

In this commit
- Import lato font and use it in the app.
- Add style to list items to match the design.

Wahaj-Ali committed:
Update country.module.css

Remove stylelint errors.

Wahaj-Ali committed:
Install "gill font" and use it in the body of the app.

In this commit
- Add gills font and use it as the body and the title font.

Wahaj-Ali committed:
Add right arrow icon in the Details page

In this commit
- Add right-arrow icon and style it.
- Remove linter errors.

Wahaj-Ali committed:
Update Dashboard styling

Wahaj-Ali committed:
Update README.md

- Add a descriptive readme file.

Wahaj-Ali committed:
Update README.md

Add live demo link.

Wahaj-Ali committed:
Update README.md

Wahaj-Ali committed:
Update README.md

Wahaj-Ali committed:
Update README.md

Wahaj-Ali committed:
Update README.md

Add Presentation video link in the readme file.

mckent05 commented on src/components/Dashboard/CountryList.js:
```diff
@@ -29,16 +25,15 @@ const CountryList = () => {
           {countryList
             ? countryList.map((country) => (
-              <li key={country.numericCode}>
+              <li key={country.numericCode} className={styles.list}>
                 <Link to={`/${country.numericCode}`} className={styles.singleCountry}>
                   <BsArrowRightCircle className={styles.direct} />
```
- According to the project requirements, your Home page should have a list of categories that could be filtered using given parameters, e.g. category name. This feature is missing in your project. Please fix this 🙏 
- **OPTIONAL** - To help improve the UI/UX of your website, I suggest aligning your countryList in the middle. Currently, it seems to be aligned to the left. Consider using `justify-content: center` on the container wrapping all countries. Please fix this 👍 👍 
<img width="947" alt="5a4aw4t" src="https://user-images.githubusercontent.com/73607512/217781861-d0ee6f58-4350-495a-8eed-d73af576ef0d.PNG">

Wahaj-Ali replied:
@mckent05 thank you for reminding me of the important project requirement. I will add the feature right away.

mckent05 reviewed (changes requested):
Hi @Wahaj-Ali, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
![almost there](https://user-images.githubusercontent.com/73607512/207265056-d056ea78-77c5-4b66-802e-694a5ed4ecf3.gif)

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Wahaj-Ali committed:
Update CountryList.js

In this commit
- Add search bar to filter through the countries as required by the reviewer.

Wahaj-Ali committed:
Update country.module.css

In this commit
- Set justify-content to 'center' for the countryList.

TadesseDev commented on src/components/Dashboard/CountryList.js:
```diff
@@ -7,20 +7,38 @@ import Country from './Country';
-    if (!countryList) {
+    if (!state) {
       dispatch(fetchApi());
     }
-  }, [dispatch, countryList]);
+  }, [dispatch, state]);
+
+  const countryList = state.filter(
+    (item) => item.name.toLowerCase().includes(search.toLowerCase()),
+  );
```
- [ ] Dear @Wahaj-Ali, You did well adding the required changes, But while doing so I notice you have introduced a bug into the project and I could not proceed to assert if the changes are implemented and your project is working fine. I have checked your project both on the `live` and `local` machines, the `dev` branch works fine but does not have the changes, and the `test` seems to be falling, I have attached the video and hope it will be helpful. 

> I also notice the exception is raised from this line. Please address this and make sure the required changes are in place.
 
### Test Brach 
https://user-images.githubusercontent.com/69077061/217799648-e99f4d11-deec-47af-b533-17d1d31fb5d0.mp4

### Dev branch 

https://user-images.githubusercontent.com/69077061/217799916-6ed325ec-295a-4657-b748-e9f60872f530.mp4

Wahaj-Ali replied:
@Tadesse-Alemayehu thank you for your suggestions. I want to mention that this PR is from the 'tests' branch to the 'dev' branch, so the dev branch does not have any latest changes. This is why the dev branch is working fine but does not contain the latest changes.

TadesseDev replied:
Thanks for reaching out @Wahaj-Ali, the reason I refer to the Dev branch is that I notice the change the previous reviewer reviewed were already on the dev branch. As a pull request submitted for review, every change should be on the new/feature branch including the initial functionalities.

TadesseDev reviewed (changes requested):
### Hi @Wahaj-Ali 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

![199733072-bdacca19-8df9-43a7-be93-3aff2f1c15c8](https://user-images.githubusercontent.com/69077061/204107907-c2532abb-b23d-4fda-88f0-1ec9c7b71971.gif)

 ## Highlights
 - [ ] Good job making most of the changes requested by the previous reviewer. 🙏

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Wahaj-Ali committed:
Update CountryList.js

- Fixed the bug for filter method.

Wahaj-Ali committed:
Updated test snapshots.

Ghiftee reviewed (approved):
## STATUS: APPROVED ♻️♻️
Hi @Wahaj-Ali, 

To highlight 👍🏽 
✔️ No linter errors
✔️ App works as expected
✔️ Tests pass

Your project is complete!  There is nothing else to say other than... it's time to merge it 🚀  
Congratulations! 🎉 

![](https://camo.githubusercontent.com/f181708ef1a4a773db8464e77479f28ed4dae506d4d8da5c19f44c8e00d31232/68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f594b646d3931584f4e4c5461772f67697068792d646f776e73697a65642d6d656469756d2e676966)

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.
